homeassistant/components/ollama/conversation.py

```diff
@@ -146,35 +217,66 @@ async def async_process(
-            )
-            return conversation.ConversationResult(
-                response=intent_response, conversation_id=conversation_id
-            )
+        # To prevent infinite loops, we limit the number of iterations
+        for _iteration in range(MAX_TOOL_ITERATIONS):
+            try:
+                response = await client.chat(
+                    model=model,
+                    # Make a copy of the messages because we mutate the list later
+                    messages=list(message_history.messages),
+                    tools=tools,
+                    stream=False,
+                    # keep_alive requires specifying unit. In this case, seconds
+                    keep_alive=f"{settings.get(CONF_KEEP_ALIVE, DEFAULT_KEEP_ALIVE)}s",
+                )
+            except (ollama.RequestError, ollama.ResponseError) as err:
+                _LOGGER.error("Unexpected error talking to Ollama server: %s", err)
+                intent_response.async_set_error(
+                    intent.IntentResponseErrorCode.UNKNOWN,
+                    f"Sorry, I had a problem talking to the Ollama server: {err}",
+                )
+                return conversation.ConversationResult(
+                    response=intent_response, conversation_id=conversation_id
+                )
 
-        response_message = response["message"]
-        message_history.messages.append(
-            ollama.Message(
-                role=response_message["role"], content=response_message["content"]
+            response_message = response["message"]
+            message_history.messages.append(
+                ollama.Message(
+                    role=response_message["role"],
+                    content=response_message.get("content"),
+                    tool_calls=response_message.get("tool_calls"),
+                )
             )
-        )
+
+            tool_calls = response_message.get("tool_calls")
+            if not tool_calls or not llm_api:
+                break
+
+            for tool_call in tool_calls:
+                tool_input = llm.ToolInput(
+                    tool_name=tool_call["function"]["name"],
+                    tool_args=tool_call["function"]["arguments"],
+                )
+                _LOGGER.debug(
+                    "Tool call: %s(%s)", tool_input.tool_name, tool_input.tool_args
+                )
+
+                try:
+                    tool_response = await llm_api.async_call_tool(tool_input)
+                except (HomeAssistantError, vol.Invalid) as e:
+                    tool_response = {"error": type(e).__name__}
+                    if str(e):
+                        tool_response["error_text"] = str(e)
+
+                _LOGGER.debug("Tool response: %s", tool_response)
+                message_history.messages.append(
+                    ollama.Message(
+                        role=MessageRole.TOOL.value,  # type: ignore[typeddict-item]
```

Comment:
We'll keep the `type: ignore[typeddict-item]` until https://github.com/ollama/ollama-python/pull/227 is merged